src/PreserveModuleNamePlugin.ts

```diff
@@ -72,90 +164,320 @@ function getPreservedModules(modules: Webpack.Module[]) {
   return new Set(
     modules.filter(m => {
       // Some modules might have [preserveModuleName] already set, see ConventionDependenciesPlugin.
-      let value = m[preserveModuleName];      
-      for (let r of m.reasons) {
-        if (!r.dependency || !r.dependency[preserveModuleName]) continue;
-        value = true;
-        let req = removeLoaders((r.dependency as ModuleDependency).request);
-        // We try to find an absolute string and set that as the module [preserveModuleName], as it's the best id.
-        if (req && !req.startsWith(".")) {
-          m[preserveModuleName] = req;
-          return true;
-        }
+      if (m[preserveModuleName]) {
+        return true;
       }
-      return !!value;
+
+      // Preserve the module if its dependencies are also preserved
+      const reasons = (m.reasons && Array.isArray(m.reasons)) ? m.reasons : [];
+      return reasons.some((reason) => Boolean(reason.dependency && reason.dependency[preserveModuleName]));
     })
   );
 }
 
-function aliasRelative(aliases: {[key: string]: string } | null, resource: string) {
-  // We consider that aliases point to local folder modules.
-  // For example: `"my-lib": "../my-lib/src"`.
-  // Basically we try to make the resource relative to the alias target,
-  // and if it works we build the id from the alias name.
-  // So file `../my-lib/src/index.js` becomes `my-lib/index.js`.
-  // Note that this only works with aliases pointing to folders, not files.
-  // To have a "default" file in the folder, the following construct works:
-  // alias: { "mod$": "src/index.js", "mod": "src" }
-  if (!aliases) return null;
-  for (let name in aliases) {
-    let root = path.resolve(aliases[name]);
-    let relative = path.relative(root, resource);
-    if (relative.startsWith("..")) continue;
-    name = name.replace(/\$$/, ""); // A trailing $ indicates an exact match in webpack
-    return relative ? name + "/" + relative : name;
-  }
-  return null;
+/**
+ *  Check if a module exists in node_modules/
+ *
+ *  @param  {Webpack.Module} module The module to check
+ *
+ *  @return {Boolean} True if it exists in node_modules/, false otherwise
+ */
+function isNodeModule(module: Webpack.Module): boolean {
```

Comment:
Coding style in this project is to infer return types, unless they can't be inferred.